### PR TITLE
universes refactor 3

### PR DESCRIPTION
**Pull request by nikomatsakis**
Some more refactorings from my universe branch. These are getting a bit more "invasive" -- they start to plumb the universe information through the canonicalization process. As of yet though I don't **believe** this branch changes our behavior in any notable way, though I'm marking the branch as `WIP` to give myself a chance to verify this.

r? @scalexm

### CODE REVIEW

File: src/librustc/infer/canonical/mod.rs [hunk 1]
```diff
@@ -104,7 +137,25 @@ pub enum CanonicalVarKind {
     Ty(CanonicalTyVarKind),
 
     /// Region variable `'?R`.
-    Region,
+    Region(ty::UniverseIndex),
+
+    /// Region variable `'?R`.
```
Comment (scalexm) [inline src/librustc/infer/canonical/mod.rs]: I don't think you wanted the exact same comment as the one above?

Comment (nikomatsakis) [reply src/librustc/infer/canonical/mod.rs]: Fixed.

**Comment by nikomatsakis** [comment]
I'm removing the WIP tag. =) Though I would *still* like to spend a bit of time looking carefully over the code. But I guess the test suite passing is a good sign I've not changed much behavior here ;)

**Comment by nikomatsakis** [comment]
There might be room though for an assert or two -- e.g., in the canonicalization code -- checking that we don't have any universes in places we don't expect them.

**Comment by scalexm** [comment]
Nothing strikes me regarding changes of behavior.

 r=me if you ever want to add those asserts

**Comment by nikomatsakis** [comment]
It seems like this failed with some sort of timeout? I suppose it's *possible* that this branch made the compiler slower, though I had not expected that.

**Comment by kennytm** [comment]
Ugh. 

@bors r- try- retry

File: src/test/ui/hrtb/hrtb-perfect-forwarding.rs [hunk 2]
```diff
@@ -29,22 +29,22 @@ impl<'a,X,F> Bar<X> for &'a mut F
...
-
-fn bar_hrtb<T>(mut t: T)
-    where T : for<'b> Bar<&'b isize>
-{
-    // OK -- `T : for<'b> Bar<&'b isize>`, and thus the impl above
-    // ensures that `&mut T : for<'b> Bar<&'b isize>`.  This is an
-    // example of a "perfect forwarding" impl.
-    bar_hrtb(&mut t);
-}
+//fn no_hrtb<'b,T>(mut t: T)
```
Comment (tmandry) [inline src/test/ui/hrtb/hrtb-perfect-forwarding.rs]: @nikomatsakis Did you mean to leave this commented out?

Comment (nikomatsakis) [reply src/test/ui/hrtb/hrtb-perfect-forwarding.rs]: Um probably not, thanks!

**Comment by nikomatsakis** [comment]
@bors p=1

@scalexm requested that I give this p=1, as they have a number of things blocked on it

**Pull request merged**